### PR TITLE
Convert NetworkController to TS

### LINKED ISSUE #18286: Convert NetworkController to TypeScript (excluding tests)
In order to be able to compare differences between the version of NetworkController in this repo and the version in the core repo before we replace this version with the core version, we ought to convert NetworkController to TypeScript.

This ticket just encompasses converting the implementation file to TypeScript, but not the tests.

### Definition of Done
- Whenever applies, any change unit tested, reviewed(approved) and documented(JSDOC at least)
- [ ] We have changelog entries for any related changes
- [ ] Any changed APIs have comprehensive inline documentation
- [ ] Any changed public APIs are well covered by unit tests
- Other items
- [ ] If a planning/research ticket, then the plan has been reviewed and approved by at least 1 team members

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

Converting this controller to TypeScript furthers the goal of getting
this whole codebase converted, of course, but it also helps in comparing
the differences between this version of the NetworkController and the
version in the `core` repo more easily, which will ultimately help us in
coalescing the two implementations.

Closes #18286.

## Manual Testing Steps

This PR should have no functional changes. All tests should pass, there should be no TypeScript errors, and extension functionality should continue to work as it does today.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: app/scripts/controllers/network/network-controller.test.js [hunk 1]
```diff
@@ -1100,7 +1100,7 @@ describe('NetworkController', () => {
     });
 
     describe('when the request for the latest block responds with null', () => {
-      it('stores null as whether the network supports EIP-1559', async () => {
+      it('persists false to state as whether the network supports EIP-1559', async () => {
```
Comment (mcmire) [inline app/scripts/controllers/network/network-controller.test.js:1103]: In order to satisfy a lint error, I've changed `_determineEIP1559Compatibility` to use `?.` when checking for the latest block. This ended up simplifying the type as well so that it always returns either true or false, never null.

File: types/@metamask/obs-store.d.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+declare module '@metamask/obs-store' {
```
Comment (mcmire) [inline types/@metamask/obs-store.d.ts:1]: This is temporary; we're using an older version of `@metamask/obs-store` where the types aren't generic enough.

File: app/scripts/controllers/network/network-controller.ts [hunk 3]
```diff
@@ -0,0 +1,1171 @@
...
+   * User-customizable details for the network.
+   */
+  rpcPrefs?: {
+    blockExplorerUrl?: string;
+  };
+  /**
+   * The ID of the network configuration associated with this provider
+   * configuration.
+   */
+  id?: NetworkConfigurationId;
```
Comment (mcmire) [inline app/scripts/controllers/network/network-controller.ts:197]: @adonesky1 Do you know if this property is actually used? I swear I saw it being set at some point but now I can't find it.

File: app/scripts/controllers/network/create-network-client.ts [hunk 4]
```diff
@@ -20,8 +20,8 @@ import {
   SafeEventEmitterProvider,
 } from '@metamask/eth-json-rpc-provider';
 import { createInfuraMiddleware } from '@metamask/eth-json-rpc-infura';
-import type { Hex } from '@metamask/utils/dist';
-import { PollingBlockTracker } from 'eth-block-tracker/dist';
+import type { Hex } from '@metamask/utils';
```
Comment (mcmire) [inline app/scripts/controllers/network/create-network-client.ts:23]: These changes are extra, but something I stumbled across. I think we ought to add `exports` to all of our projects to prevent this from happening. I can file an issue in the module template.

File: package.json [hunk 5]
```diff
@@ -378,6 +378,7 @@
     "@metamask/eslint-config-typescript": "^9.0.1",
     "@metamask/forwarder": "^1.1.0",
     "@metamask/phishing-warning": "^2.1.0",
+    "@metamask/safe-event-emitter": "^2.0.0",
```
Comment (Gudahtt) [inline package.json:381]: The older `safe-event-emitter` package is in `dependencies` still. We should be able to update that rather than adding another copy. I've done that in a separate PR: https://github.com/MetaMask/metamask-extension/pull/18419

File: app/scripts/controllers/network/create-network-client.ts (hunk 4 above)
Comment (mcmire) [reply app/scripts/controllers/network/create-network-client.ts:23]: Well, I guess we don't need these changes here though. I can extract this to a new PR.

File: app/scripts/controllers/network/network-controller.ts [hunk 6]
```diff
@@ -0,0 +1,1171 @@
...
+type Block = {
+  baseFeePerGas?: unknown;
+};
+
+/**
+ * Encodes both whether or not a provider is configured for an Infura network or
+ * a non-Infura network; if for an Infura network, then which network; and if
+ * for a non-Infura network, then whether the network is being operated locally
+ * or remotely. Primarily used to build the network client and check the
+ * availability of a network.
```
Comment (Gudahtt) [inline app/scripts/controllers/network/network-controller.ts:60]: Nit: This sentence was confusing. Maybe we could simplify it somehow? The structure of "Both [first thing thing]; if ... then [second thing]; if ... then [second thing]" was hard to follow and doesn't seem grammatically correct.

File: app/scripts/controllers/network/network-controller.ts [hunk 7]
```diff
@@ -0,0 +1,1171 @@
...
+
+/**
+ * The messenger that the NetworkController uses to publish events.
+ */
+type NetworkControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  never,
+  NetworkControllerEvent,
+  never,
+  NetworkControllerEventType
```
Comment (Gudahtt) [inline app/scripts/controllers/network/network-controller.ts:157]: We don't need to list the internal events here, they're allowed by default. This generic parameter is just meant for _external_ events that are allowed.

File: app/scripts/controllers/network/network-controller.ts (hunk 3 above)
Comment (Gudahtt) [reply app/scripts/controllers/network/network-controller.ts:197]: It has definitely never been used as "provider configuration". Our middleware stack doesn't use this ID for anything.

**Comment by Gudahtt** [comment]
Looks good! Just finished a first pass and the only feedback I had was regarding comments.

All of the related PRs have been merged, so it should be possible to rebase this on `develop` and make this ready for review now.

File: app/scripts/controllers/network/network-controller.ts (hunk 6 above)
Comment (mcmire) [reply app/scripts/controllers/network/network-controller.ts]: You're right, this was confusing. Hopefully this is better: [84140af](https://github.com/MetaMask/metamask-extension/pull/18358/commits/84140afaf0b2ea0dd70a6bc14fa9cfcbad1c11cb)

File: app/scripts/controllers/network/network-controller.ts [hunk 8]
```diff
@@ -0,0 +1,1171 @@
...
+};
+
+/**
+ * Type guard for determining whether the given value is an error object with a
+ * `code` property, such as an instance of Error.
+ *
+ * TODO: Move this to @metamask/utils
+ *
+ * @param error - The object to check.
+ * @returns True or false, depending on the result.
```
[the remark this answers is not in the document]
Comment (mcmire) [reply app/scripts/controllers/network/network-controller.ts]: Good call. Fixed in [c890d90](https://github.com/MetaMask/metamask-extension/pull/18358/commits/c890d904f3a9e1e938aff4cac5132df7329352f3).

File: app/scripts/controllers/network/network-controller.ts [hunk 9]
```diff
@@ -0,0 +1,1171 @@
...
+/**
+ * Asserts that the given value is a network ID, i.e., that it is a decimal
+ * number represented as a string.
+ *
+ * @param value - The value to check.
+ */
+function assertNetworkId(value: any): asserts value is NetworkId {
+  assert(
+    /^\d+$/u.test(value) && !Number.isNaN(Number(value)),
+    'value cannot be converted to a number',
```
[the remark this answers is not in the document]
Comment (mcmire) [reply app/scripts/controllers/network/network-controller.ts]: Makes sense. Updated in [f243281](https://github.com/MetaMask/metamask-extension/pull/18358/commits/f2432813788972a7004aa1b8fc5a4d95404b3eb4).

File: app/scripts/controllers/network/network-controller.ts (hunk 3 above)
Comment (mcmire) [reply app/scripts/controllers/network/network-controller.ts]: Okay, I've removed this for now: [de0c195](https://github.com/MetaMask/metamask-extension/pull/18358/commits/de0c195d6a1633f3af189be91b548e5e669f9af2). Happy to add this back later though if we need it.

File: app/scripts/controllers/network/create-network-client.ts (hunk 4 above)
Comment (mcmire) [reply app/scripts/controllers/network/create-network-client.ts]: Removed for now: [88811d2](https://github.com/MetaMask/metamask-extension/pull/18358/commits/88811d2e189b70e0bd3b50c427486acb201e9d4f)

File: app/scripts/controllers/network/network-controller.ts (hunk 6 above)
Comment (Gudahtt) [reply app/scripts/controllers/network/network-controller.ts]: That looks great! I'm glad I didn't suggest an alternative, this is better than what I would have come up with.

File: types/@metamask/obs-store.d.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import { ObservableStore } from '@metamask/obs-store';
+
+declare module '@metamask/obs-store' {
+  export class ComposedStore<
```
Comment (mcmire) [inline types/@metamask/obs-store.d.ts:4]: Overriding this is still necessary because the types for ComposedStore assume that each store is an object.

Comment (mcmire) [reply types/@metamask/obs-store.d.ts:4]: @Gudahtt Should I try to update `@metamask/obs-store` with these changes?

Comment (Gudahtt) [reply types/@metamask/obs-store.d.ts:4]: Ah interesting. Yeah I guess so? I see that in the ObservableStore class they made the same mistake in assuming the state was an object, but then fixed it by using generics instead. Applying the same fix to the composable store would make sense.

Comment (mcmire) [reply types/@metamask/obs-store.d.ts:4]: Okay, PRs created here: https://github.com/MetaMask/obs-store/pull/79, https://github.com/MetaMask/obs-store/pull/80

**Comment by mcmire** [comment]
@Gudahtt I want to get a new release for `@metamask/obs-store` out first and then I'll open this for review.

File: types/@metamask/obs-store.d.ts (hunk 2 above)
Comment (mcmire) [reply types/@metamask/obs-store.d.ts]: With the new release, I was able to remove this file.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Review by Gudahtt — APPROVED** [review]

**Review by legobeat — APPROVED** [review]
